// @sphereon/pex/dist/browser/lib/validation/core/presentationSubmissionSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PresentationSubmissionSchema = void 0;
class PresentationSubmissionSchema {
    //TODO: pass it with a config file
    static getPresentationSubmissionSchema() {
        return {
            $schema: 'http://json-schema.org/draft-07/schema#',
            title: 'Presentation Submission',
            type: 'object',
            properties: {
                presentation_submission: {
                    type: 'object',
                    properties: {
                        id: {
                            type: 'string',
                        },
                        definition_id: {
                            type: 'string',
                        },
                        descriptor_map: {
                            type: 'array',
                            items: {
                                $ref: '#/definitions/descriptor',
                            },
                        },
                    },
                    required: ['id', 'definition_id', 'descriptor_map'],
                    additionalProperties: false,
                },
            },
            definitions: {
                descriptor: {
                    type: 'object',
                    properties: {
                        id: {
                            type: 'string',
                        },
                        path: {
                            type: 'string',
                        },
                        path_nested: {
                            type: 'object',
                            $ref: '#/definitions/descriptor',
                        },
                        format: {
                            type: 'string',
                            enum: ['jwt', 'jwt_vc', 'jwt_vc_json', 'jwt_vp', 'ldp', 'ldp_vc', 'ldp_vp'],
                        },
                    },
                    required: ['id', 'path', 'format'],
                    additionalProperties: false,
                },
            },
            required: ['presentation_submission'],
            additionalProperties: false,
        };
    }
}
exports.PresentationSubmissionSchema = PresentationSubmissionSchema;
