// @sphereon/ssi-types/dist/types/vc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusListDriverType = exports.StatusListCredentialIdMode = exports.StatusListType = exports.JWT_PROOF_TYPE_2020 = exports.DocumentFormat = exports.OriginalType = void 0;
var OriginalType;
(function (OriginalType) {
    OriginalType["JSONLD"] = "json-ld";
    OriginalType["JWT_ENCODED"] = "jwt-encoded";
    OriginalType["JWT_DECODED"] = "jwt-decoded";
})(OriginalType = exports.OriginalType || (exports.OriginalType = {}));
var DocumentFormat;
(function (DocumentFormat) {
    DocumentFormat[DocumentFormat["JWT"] = 0] = "JWT";
    DocumentFormat[DocumentFormat["JSONLD"] = 1] = "JSONLD";
    DocumentFormat[DocumentFormat["EIP712"] = 2] = "EIP712";
})(DocumentFormat = exports.DocumentFormat || (exports.DocumentFormat = {}));
exports.JWT_PROOF_TYPE_2020 = 'JwtProof2020';
var StatusListType;
(function (StatusListType) {
    StatusListType["StatusList2021"] = "StatusList2021";
})(StatusListType = exports.StatusListType || (exports.StatusListType = {}));
var StatusListCredentialIdMode;
(function (StatusListCredentialIdMode) {
    StatusListCredentialIdMode["ISSUANCE"] = "ISSUANCE";
    StatusListCredentialIdMode["PERSISTENCE"] = "PERSISTENCE";
    StatusListCredentialIdMode["NEVER"] = "NEVER";
})(StatusListCredentialIdMode = exports.StatusListCredentialIdMode || (exports.StatusListCredentialIdMode = {}));
var StatusListDriverType;
(function (StatusListDriverType) {
    StatusListDriverType["AGENT_TYPEORM"] = "agent_typeorm";
    StatusListDriverType["AGENT_KV_STORE"] = "agent_kv_store";
    StatusListDriverType["GITHUB"] = "github";
    StatusListDriverType["AGENT_FILESYSTEM"] = "agent_filesystem";
})(StatusListDriverType = exports.StatusListDriverType || (exports.StatusListDriverType = {}));
